### PR TITLE
Supports mutli-turn REVOLUTE joints

### LINKED ISSUE #2: [gazebo_ros_control] Supports multi-turn joints when using PID mode
**Summery**
- For REVOLUTE type joints, joints with a range of motion greater than one revolution do not operate properly. (Only when PID is enabled)
- The following programs are causing the problem
https://github.com/sbgisen/gazebo_ros_pkgs/blob/b0e3d71938e4cd4c80a2ec3ff7187aac49a9f62a/gazebo_ros_control/src/default_robot_hw_sim.cpp#L328-L346
- REVOLUTE joint has exception handling to keep the target angle within the range of motion, but this includes a function that does not work correctly if the range of motion is more than one revolution.
- The CONTINUOUS joint can handle more than one rotation, but the algorithm is designed to reverse the rotation in an attempt to take a shortcut when the target angle deviates more than half a revolution from the current angle.
- If the "default" in the case statement is used as is, the operation will correspond to a multi-turn joint.

### PR DESCRIPTION
**Pull request by nyxrobotics**
## Summary
Added exception handling to control REVOLUTE type joints with a range of motion greater than one revolution

- fix #2

### CODE REVIEW

**Comment by jsupratman13** [comment]
I haven't fully checked the code properly but I was wondering if the `angles::shortest_angular_distance_with_large_limits` function will suffice?

**Comment by nyxrobotics** [comment]
https://github.com/ros/angles/blob/fe974f2d84b719d3aa0593a4bf633183d75ba213/angles/include/angles/angles.h#L222-L269

**Comment by nyxrobotics** [comment]
That function looks like an expression that rotates backwards if the target angle jumps more than half a turn from the current angle, for the same reason as ```angles::shortest_angular_distance```

File: gazebo_ros_control/src/default_robot_hw_sim.cpp [hunk 1]
```diff
@@ -331,11 +331,19 @@ void DefaultRobotHWSim::writeSim(ros::Time time, ros::Duration period)
           switch (joint_types_[j])
           {
             case urdf::Joint::REVOLUTE:
-              angles::shortest_angular_distance_with_limits(joint_position_[j],
-                                                            joint_position_command_[j],
-                                                            joint_lower_limits_[j],
-                                                            joint_upper_limits_[j],
-                                                            error);
+              if(joint_upper_limits_[j] - joint_lower_limits_[j] > 2.0 * M_PI)
```
Comment (jsupratman13) [inline gazebo_ros_control/src/default_robot_hw_sim.cpp:334]: Use absolute value to determine if the limits (the rotatable revolutions) exceeds 2*pi

**Review by jsupratman13 — CHANGES_REQUESTED** [review]
This looks more like a safety net than multi-turn support. If this is the PR's intention then I have no problem with the code.

I am a little uneasy about using the forked package to simulate gazebo.

**Comment by nyxrobotics** [comment]
Then we will use this branch as it is, shall we put out a PR for the main repository?

File: gazebo_ros_control/src/default_robot_hw_sim.cpp (hunk 1 above)
Comment (nyxrobotics) [reply gazebo_ros_control/src/default_robot_hw_sim.cpp:334]: Do you mean `joint_upper_limits_[j]` can be smaller than `joint_lower_limits_[j]` ?

Comment (jsupratman13) [reply gazebo_ros_control/src/default_robot_hw_sim.cpp:334]: oops i forgot about the upper > lower limit
this should be ok

**Comment by jsupratman13** [comment]
> Then we will use this branch as it is, shall we put out a PR for the main repository?

I'm good with that too. If you want, we could also set up an issue on the main repository first and see if there is a response from the dev (though I think someone else did address this issue but no response has been made to it)

I'm more concerned about the situations that lead to this PR. Was there a URDF, in our repositories, that exceeds the 1 revolution range?

**Review by jsupratman13 — APPROVED** [review]

**Comment by jsupratman13** [comment]
please include CATKIN_IGNORE files on unchanged packages

**Review by Tacha-S — APPROVED** [review]

**Comment by tokoro10g** [comment]
@nyxrobotics 
Hi, would you mind sending this patch to the upstream? We are facing the same issue, and I believe that it should be included in the original repository. What would you say?
If you would like me to handle it, I would be happy to do that.

**Comment by nyxrobotics** [comment]
@tokoro10g 
There was also an [issue](https://github.com/ros-simulation/gazebo_ros_pkgs/issues/812) in the official repository that was not gone through, so I forked and modified it.
I have now commented on the issue.
If the issue becomes active, there may be a chance. If you have an opinion, please comment.

**Pull request merged**